enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/ArtifactUtils.java

```diff
@@ -18,70 +18,188 @@
 
 /**
  *
  * @author Robert Scholte
  * @since 3.0.0
  */
 public final class ArtifactUtils {
-    private ArtifactUtils() {}
 
-    public static Set<Artifact> getAllDescendants(DependencyNode node) {
-        Set<Artifact> children = null;
-        if (node.getChildren() != null) {
-            children = new HashSet<>();
-            for (DependencyNode depNode : node.getChildren()) {
-                children.add(depNode.getArtifact());
-                Set<Artifact> subNodes = getAllDescendants(depNode);
-                if (subNodes != null) {
-                    children.addAll(subNodes);
-                }
+    /**
+     * Converts {@link DependencyNode} to {@link Artifact}; in comparison
+     * to {@link RepositoryUtils#toArtifact(org.eclipse.aether.artifact.Artifact)}, this method
+     * assigns {@link Artifact#getScope()} and {@link Artifact#isOptional()} based on
+     * the dependency information from the node.
+     *
+     * @param node {@link DependencyNode} to convert to {@link Artifact}
+     * @return target artifact
+     */
+    public static Artifact toArtifact(DependencyNode node) {
+        Artifact artifact = RepositoryUtils.toArtifact(node.getArtifact());
+        ofNullable(node.getDependency()).ifPresent(dependency -> {
+            ofNullable(dependency.getScope()).ifPresent(artifact::setScope);
+            artifact.setOptional(dependency.isOptional());
+        });
+        return artifact;
+    }
+
+    /**
+     * Retrieves the {@link DependencyNode} instance containing the result of the transitive dependency
+     * for the current {@link MavenProject}.
+     *
+     * @param helper (may not be null) an instance of the {@link EnforcerRuleHelper} class
+     * @param selectors zero or more {@link DependencySelector} instances
+     * @return a Dependency Node which is the root of the project's dependency tree
+     * @throws EnforcerRuleException thrown if the lookup fails
+     */
+    public static DependencyNode resolveTransitiveDependencies(
+            EnforcerRuleHelper helper, DependencySelector... selectors) throws EnforcerRuleException {
+        try {
+            RepositorySystem repositorySystem = helper.getComponent(RepositorySystem.class);
+            MavenSession session = (MavenSession) helper.evaluate("${session}");
+            MavenProject project = session.getCurrentProject();
+            ArtifactTypeRegistry artifactTypeRegistry =
+                    session.getRepositorySession().getArtifactTypeRegistry();
+
+            DefaultRepositorySystemSession repositorySystemSession =
+                    new DefaultRepositorySystemSession(session.getRepositorySession());
+            repositorySystemSession.setDependencyGraphTransformer(new ConflictResolver(
+                    new NearestVersionSelector(),
+                    new JavaScopeSelector(),
+                    new SimpleOptionalitySelector(),
+                    new JavaScopeDeriver()));
```

Comment:
This is default settings provided by Maven - why we need to change?

https://github.com/apache/maven/blob/master/maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/MavenRepositorySystemUtils.java

Reply:
Ok